cbor2/decoder.py

```diff
@@ -304,13 +321,14 @@ class CBORDecoder(object):
         The return value is substituted for the dict in the deserialized output.
     """
 
-    __slots__ = ('fp', 'tag_hook', 'object_hook', '_shareables')
+    __slots__ = ('fp', 'tag_hook', 'object_hook', '_shareables', '_immutable')
 
     def __init__(self, fp, tag_hook=None, object_hook=None):
         self.fp = fp
         self.tag_hook = tag_hook
         self.object_hook = object_hook
         self._shareables = []
+        self._immutable = False
```

Comment:
My previous comment got lost somewhere, but if decoders must access both `key_decoder` and `_immutable`, why is one public and one private? Shouldn't both be public if third party decoders are allowed?

Reply:
Hm, I guess it would make sense to make that flag generally global, but you never want to set it to False until the last context manager has exited.

For instance, the object_hook and tag_hook may need to know if it's being passed something contained within a map.

Reply:
How about this:

```python
@property
def immutable(self):
    return self._immutable
```

Without the corresponding setter method, then the internal flag would only be set by the context manager but it would be readable as:

```python
if decoder.immutable:
    return tuple(value)
```

Reply:
Works for me.